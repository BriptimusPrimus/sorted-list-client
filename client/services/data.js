import config from '../../config';

const apiEndpoint = config.data_service.api.protocol + '://'
  + config.data_service.api.host + ':'
  + config.data_service.api.port + '/'
  + config.data_service.api.root + '/';

function getRequest(endpoint) {
  function makeRequest(resolve, reject) {
    const requestConf = {
      method: 'GET',
      mode: 'cors',
      headers: new Headers({
        'Content-Type': 'application/json'
      })
    }

    fetch(endpoint, requestConf)
      .then(function(response) {
        if (response.status !== 200) {
          return {
            error: 'Service Failure',
            status: response.status
          }
        }
        return response.json();
      })
      .then(function(data) {
        if(data.error) {
          reject(data);
        } else {
          resolve(data);
        }
      })
      .catch(function(reason) {
        reject({
          error: 'Service Failure'
        });
      });
  }

  return new Promise(function(resolve, reject) {
    makeRequest(resolve, reject);
  });
}

function postRequest(enpoint, data) {
  function makeRequest(resolve, reject) {
    const requestConf = {
      method: 'POST',
      mode: 'cors',
      headers: new Headers({
        'Content-Type': 'application/json'
      }),
      body: JSON.stringify(data)
    }

    fetch(endpoint, requestConf)
      .then(function(response) {
        if (response.status !== 200) {
          return {
            error: 'Service Failure',
            status: response.status
          }
        }        
        return response.json();
      })
      .then(function(data) {
        if(data.error) {
          reject(data);
        } else {
          resolve(data);
        }
      })
      .catch(function(reason) {
        reject({
          error: 'Service Failure'
        });
      });
  }

  return new Promise(function(resolve, reject) {
    makeRequest(resolve, reject);
  });

}

function getEmployees() {
  const endpoint = apiEndpoint
    + config.data_service.api.employees_data +
    '?count=' + config.data_service.defaulCount;

  return getRequest(endpoint);
}

export default {
  getEmployees
}