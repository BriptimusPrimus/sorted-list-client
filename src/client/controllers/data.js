import dataService from '../services/data';
import {sortByColumn} from '../actions';

function sortBy(dispatchFn, column) {
  const fetchOpts = {
    sortBy: store.getState().sortBy
  };  
  
  dispatchFn(sortByColumn(column));
  dataService.getEmployees(fetchOpts)
    .then(function fullfilled(data) {
      store.dispatch(receiveData(data));
    })
    .catch(function rejected(reason) {
      console.log('response error:', reason);
    })
}

export default {
  sortBy
}