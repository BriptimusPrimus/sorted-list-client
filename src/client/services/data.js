import fetch from 'isomorphic-fetch';
import querystring from 'querystring';
import config from '../../../config';

const apiEndpoint = `${config.data_service.api.protocol}://${config.data_service.api.host}:${config.data_service.api.port}/${config.data_service.api.root}/`;

const getRequest = async function getRequest(endpoint) {
  const requestConf = {
    method: 'GET',
    mode: 'cors',
    headers: new Headers({
      'Content-Type': 'application/json'
    })
  };

  const response = await fetch(endpoint, requestConf);

  if (response.status === 404) {
    return {
      notFound: true
    };
  }

  if (response.status !== 200) {
    return {
      error: 'Service Failure',
      status: response.status
    };
  }

  const data = await response.json();
  if (data.error) {
    throw new Error(data);
  }

  return data;
};

const getEmployees = function getEmployees(options = {}) {
  const sortBy = options.sortBy || {
    column: 'codeNumber'
  };

  const params = {
    count: options.count || config.data_service.defaulCount,
    sort: sortBy.column,
    order: sortBy.orderDesc ? 'DESC' : 'ASC'
  };

  const endpoint = `${
    apiEndpoint + config.data_service.api.employees_data
  }?${querystring.stringify(params)}`;

  return getRequest(endpoint);
};

const getCustomers = function getCustomers(options = {}) {
  const sortBy = options.sortBy || {
    column: 'id'
  };

  const params = {
    count: options.count || config.data_service.defaulCount,
    sort: sortBy.column,
    order: sortBy.orderDesc ? 'DESC' : 'ASC'
  };

  const endpoint = `${
    apiEndpoint + config.data_service.api.customers_data
  }?${querystring.stringify(params)}`;

  return getRequest(endpoint);
};

const getCustomerDetails = function getCustomerDetails({ customerId } = {}) {
  const endpoint = `${
    apiEndpoint + config.data_service.api.customer_details
  }/${customerId}`;

  return getRequest(endpoint);
};

const getDisplayDetails = function getDisplayDetails({ customerId } = {}) {
  const endpoint = `${
    apiEndpoint + config.data_service.api.display_details
  }?id=${customerId}`;

  return getRequest(endpoint);
};

export { getEmployees, getCustomers, getCustomerDetails, getDisplayDetails };
