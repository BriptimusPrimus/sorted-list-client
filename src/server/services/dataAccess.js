/*
 * Data Access Service Abstraction
 */

const implementationModule = require('../../../config').dal.module;

// eslint-disable-next-line import/no-dynamic-require
const dal = require(`./${implementationModule}`);

const getCustomerDetails = async function getCustomerDetails(id) {
  try {
    const data = await dal.getCustomer(id);

    // Customer not exist
    if (data === undefined) {
      return {
        success: true,
        noDataFound: true
      };
    }

    return {
      success: true,
      data
    };
  } catch (err) {
    return {
      success: false,
      reason: (err && err.message) || err || ''
    };
  }
};

const getCustomers = async function getCustomers(opts) {
  const params = {
    count: Number.isNaN(Number(opts.count)) ? 20 : Number(opts.count),
    page:
      Number.isNaN(Number(opts.page)) || opts.page < 1 ? 1 : Number(opts.page),
    sortBy: opts.sortBy,
    order: opts.order || 'ASC',
    filters: opts.filters || {}
  };

  try {
    const data = await dal.getCustomers(params);
    return {
      success: true,
      data
    };
  } catch (err) {
    return {
      success: false,
      reason: (err && err.message) || err || ''
    };
  }
};

const getEmployees = async function getEmployees(opts) {
  const params = {
    count: Number.isNaN(Number(opts.count)) ? 20 : Number(opts.count),
    page:
      Number.isNaN(Number(opts.page)) || opts.page < 1 ? 1 : Number(opts.page),
    sortBy: opts.sortBy,
    order: opts.order || 'ASC',
    filters: opts.filters || {}
  };

  try {
    const data = await dal.getEmployees(params);
    return {
      success: true,
      data
    };
  } catch (err) {
    return {
      success: false,
      reason: (err && err.message) || err || ''
    };
  }
};

// This object implements the bridge interface:
// interface: {
//   getCustomerDetails: async function(customerId){}
//   getCustomers: async function(options){}
//   getEmployees: async function(options){}
// }
module.exports = {
  getCustomerDetails,
  getCustomers,
  getEmployees
};
