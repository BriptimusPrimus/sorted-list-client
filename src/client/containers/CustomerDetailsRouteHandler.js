import React from 'react';
import PropTypes from 'prop-types';
import { receiveData, receiveDataError } from '../actions';
import { getCustomerDetails } from '../services/data';
import CustomerDetailsContainer from './CustomerDetailsContainer';
import withUniversalLoad from '../lib/withUniversalLoad';

const CustomerDetailsWithUniversalLoad = withUniversalLoad(
  CustomerDetailsContainer,
  {
    dataSource: getCustomerDetails,
    dataSourceParams: {
      customerId: {
        inParams: true
      }
    },
    successAction: receiveData,
    failureAction: receiveDataError
  }
);

const CustomerDetailsRouteHandler = ({ initialState }) => {
  return <CustomerDetailsWithUniversalLoad initialState={initialState} />;
};

// Override loadData
CustomerDetailsRouteHandler.loadData = async function preloadData(params) {
  const data = await CustomerDetailsWithUniversalLoad.loadData(params);
  return {
    data
  };
};

export default CustomerDetailsRouteHandler;

CustomerDetailsRouteHandler.propTypes = {
  initialState: PropTypes.shape({
    data: PropTypes.shape({
      customerDetails: PropTypes.shape({
        id: PropTypes.number,
        firstName: PropTypes.string,
        surname: PropTypes.string,
        email: PropTypes.string,
        addressId: PropTypes.number,
        isActive: PropTypes.bool,
        createDate: PropTypes.string,
        lastUpdate: PropTypes.string,
        address: PropTypes.string,
        address2: PropTypes.string,
        addressLastUpdate: PropTypes.string,
        city: PropTypes.string,
        country: PropTypes.string,
        district: PropTypes.string,
        phone: PropTypes.string,
        postal_code: PropTypes.string
      })
    })
  })
};

CustomerDetailsRouteHandler.defaultProps = {
  initialState: {
    data: {}
  }
};
