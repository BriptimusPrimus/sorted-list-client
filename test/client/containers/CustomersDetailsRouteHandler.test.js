import React from 'react';
import { render, cleanup } from '@testing-library/react';
import CustomerDetailsRouteHandler from '../../../src/client/containers/CustomerDetailsRouteHandler';
import * as mockService from '../../../src/client/services/data';
import { customerDetails } from '../utils/customerDataMocks';

jest.mock('react-router-dom', () => {
  return {
    useParams: () => ({
      customerId: 42
    }),
    Link: ({ children }) => {
      return children;
    }
  };
});
jest.mock('../../../src/client/services/data');

const getInitialState = () => ({
  customerId: 42,
  data: {
    customerDetails
  }
});

describe('CustomerDetailsRouteHandler tests', () => {
  const initialState = getInitialState();

  beforeEach(() => {
    jest.clearAllMocks();
    mockService.getCustomerDetails.mockImplementation(async () => {
      return initialState.data;
    });
  });

  afterEach(cleanup);

  test('Renders', () => {
    const { container } = render(
      <CustomerDetailsRouteHandler initialState={initialState} />
    );
    expect(container.firstChild).not.toBeNull();
    expect(container.firstChild).toMatchSnapshot();
  });

  test('Customer not found', () => {
    const localState = {
      customerId: 999,
      data: {
        notFound: true
      }
    };
    const { queryByText } = render(
      <CustomerDetailsRouteHandler initialState={localState} />
    );
    expect(queryByText('Data not found')).not.toBeNull();
  });

  test('Static load data', async () => {
    const result = await CustomerDetailsRouteHandler.loadData();
    expect(result).toEqual({
      data: initialState.data
    });
  });
});
