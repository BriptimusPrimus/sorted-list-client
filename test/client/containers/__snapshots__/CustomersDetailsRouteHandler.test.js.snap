// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CustomerDetailsRouteHandler tests Renders 1`] = `
<div
  class="panel panel-default row"
>
  <div
    class="col-xs-0 col-sm-1 col-md-2"
  />
  <section
    class="col-xs-12 col-sm-10 col-md-8"
  >
    Go back to list
    <h2
      class="text-primary text-center"
    >
      Customer Details
    </h2>
    <div
      class="well row"
    >
      <table
        class="table table-striped"
      >
        <tbody>
          <tr>
            <td
              class="details-key"
            >
              Name:
            </td>
            <td>
              CAROLYN
               
              PEREZ
            </td>
          </tr>
          <tr>
            <td
              class="details-key"
            >
              Email:
            </td>
            <td>
              CAROLYN.PEREZ@sakilacustomer.org
            </td>
          </tr>
          <tr>
            <td
              class="details-key"
            >
              Phone number:
            </td>
            <td>
              471675840679
            </td>
          </tr>
          <tr>
            <td
              class="details-key"
            >
              Address:
            </td>
            <td>
              1632 Bislig Avenue
            </td>
          </tr>
          <tr>
            <td
              class="details-key"
            >
              City:
            </td>
            <td>
              Pak Kret
            </td>
          </tr>
          <tr>
            <td
              class="details-key"
            >
              District:
            </td>
            <td>
              Nonthaburi
            </td>
          </tr>
          <tr>
            <td
              class="details-key"
            >
              Country:
            </td>
            <td>
              Thailand
            </td>
          </tr>
          <tr>
            <td
              class="details-key"
            >
              Postal code:
            </td>
            <td>
              61117
            </td>
          </tr>
        </tbody>
      </table>
    </div>
  </section>
  <div
    class="col-xs-0 col-sm-1 col-md-2"
  />
</div>
`;
